src/overlay/Peer.cpp

```diff
@@ -835,8 +1001,34 @@ Peer::recvMessage(StellarMessage const& stellarMsg)
         cat = "MISC";
     }
 
-    std::weak_ptr<Peer> weak = shared_from_this();
-    auto msgTracker = std::make_shared<MsgCapacityTracker>(weak, stellarMsg);
+    auto queueName = isAuthenticated(guard) ? cat : AUTH_ACTION_QUEUE;
+    type = isAuthenticated(guard) ? type : Scheduler::ActionType::NORMAL_ACTION;
+    mAppConnector.postOnMainThread(
+        [self = shared_from_this(), msgTracker]() {
+            self->recvMessage(msgTracker);
+        },
+        std::move(queueName), type);
+}
+
+void
+Peer::recvMessage(std::shared_ptr<MsgCapacityTracker> msgTracker)
```

Comment:
I agree that'd be ideal. Unfortunately there was an underlying asio issue in passing data between IO services, where it needed to be copyable. I also had to remove the destructor, because asio doesn't provide any guarantees regarding when referenced lambdas are destroyed, so I didn't want to rely on RAII here.

Reply:
Ok, I experimented a bit more, and ended up still using shared_ptr, but moved it into the lambda to ensure that background thread isn't holding any references. We still can't use unique_ptr unfortunately, because std::function doesn't allow its callable to be move-only, so I added a comment about this. I think this is still safer than manually calling `finish`.